algebra-core/src/fields/macros.rs

```diff
@@ -431,27 +431,72 @@ macro_rules! impl_Fp {
     }
 }
 
+
+// The no carry optimisation requires last bit to be zero
+// and all remaining bits not set
+// https://hackmd.io/@zkteam/modular_multiplication
+// It is slightly faster
 macro_rules! impl_field_mul_assign {
     ($limbs:expr) => {
         #[inline]
         #[unroll_for_loops]
         fn mul_assign(&mut self, other: &Self) {
-            let mut r = [0u64; $limbs];
-            let mut carry1 = 0u64;
-            let mut carry2 = 0u64;
+            let mut no_carry = true;
+            if P::MODULUS.0[$limbs-1] >> 63 != 0 {
+                no_carry = false;
+            } else if ((P::MODULUS.0[$limbs-1] << 1) + 1) & !(0u64) == 0 {
+                let mut all_bits_set = true;
+                for i in 1..$limbs {
+                    if P::MODULUS.0[$limbs-1-i] & !(0u64) == 0 {
+                        all_bits_set &= true;
+                    } else { all_bits_set = false; }
+                }
+                no_carry = !all_bits_set;
+            }
```

Comment:
Just to ensure that it always gets compiled away, maybe we should wrap the logic in a `const fn`? In this case maybe we can move the `const fn` out of the function. I also included some simplifications of the logic, but you should double-check that these  simplifications are actually correct.

```suggestion
			#[inline(always)]
			#[unroll_for_loops]
			const fn can_use_no_carry_opt(modulus: [u64; $limbs]) -> bool {
	            if (P::MODULUS.0[$limbs - 1] >> 63) != 0 {
	                false
	            } else if (!P::MODULUS.0[$limbs - 1]) == (1 << 63) {
	            	// after flipping the bits of the last limb, every bit 
	            	// except the first was set
	            	
	                let mut all_bits_set = true;
	                for i in 1..$limbs {
	                    // the limb is equal to all-ones.
	                    all_bits_set &= (P::MODULUS.0[$limbs - i - 1] == !0u64);
	                }
	                !all_bits_set
	            } else {
	            	true
	            }
			}
			
			let no_carry = can_use_no_carry_opt(P::MODULUS.0);
```

Reply:
Unfortunately, due to the nature of `unroll_for_loops`, loops and functions containing loops are not seen as constant by the compiler. I have opened in issue https://gitlab.com/elrnv/unroll/-/issues/2 but it won't be moving for a while at least, I suspect. For now, I suggest to believe in the compiler. Constant propogation is the most basic compiler optimisation :)